Call element.set immediately when a property of an object inside a list is updated

gilberto-torrezan opened the pull request:
Fix #4064

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/flow/4152)
<!-- Reviewable:end -->

Legioth commented:
Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---

*[flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java, line 182 at r1](https://reviewable.io/reviews/vaadin/flow/4152#-LD6oWUxAgdIbrfJsiJ8:-LD6oWUxAgdIbrfJsiJ9:bakte6v) ([raw file](https://github.com/vaadin/flow/blob/5f3f7f1c1e39c1a822a4b994956e365a271de488/flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java#L182)):*
> ```Java
>     /**
>      * Sets a property value on objects inside lists defined in a TemplateModel.
>      * It's a NO-OP in case the property is not part of an object inside a list.
> ```

Seems like it's also a no-op in any case where the list isn't directly a child of the element?

---

*[flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java, line 227 at r1](https://reviewable.io/reviews/vaadin/flow/4152#-LD6mrFn1fs5ZSyr7tuR:-LD6mrFn1fs5ZSyr7tuS:bvx6uwf) ([raw file](https://github.com/vaadin/flow/blob/5f3f7f1c1e39c1a822a4b994956e365a271de488/flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java#L227)):*
> ```Java
>             String parentPropertyName, MapProperty nodeProperty, Object value) {
> 
>         UpdatableModelProperties updatableModelProperties = parent
> ```

From my understanding, `UpdatableModelProperties` is about client -> server updates, whereas we're currently processing server -> client updates? What's the connection?

---

*[flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java, line 259 at r1](https://reviewable.io/reviews/vaadin/flow/4152#-LD6n_qU0DKjfbqGmnG-:-LD6n_qU0DKjfbqGmnG0:b-k6i5z3) ([raw file](https://github.com/vaadin/flow/blob/5f3f7f1c1e39c1a822a4b994956e365a271de488/flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java#L259)):*
> ```Java
>                 .append(nodeProperty.getName());
> 
>         NativeFunction setValueFunction = NativeFunction.create("path", "value",
> ```

Also (though this also applies to the original code in `SimpleElementBindingStrategy.setSubProperties`), I don't understand why it's necessary to parse the same static JS snippet every time we get here instead of calling a JSNI method that does the same thing.

---

*[flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java, line 260 at r1](https://reviewable.io/reviews/vaadin/flow/4152#-LD6o8Sh2ykvdHI2rTyF:-LD6o8Sh2ykvdHI2rTyG:bikyiy1) ([raw file](https://github.com/vaadin/flow/blob/5f3f7f1c1e39c1a822a4b994956e365a271de488/flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java#L260)):*
> ```Java
> 
>         NativeFunction setValueFunction = NativeFunction.create("path", "value",
>                 "this.set(path, value)");
> ```

Is there any guarantee that the target element has a `set` function (i.e. that it's an upgraded polymer element)? There was a fix yesterday for the same case with `SimpleElementBindingStrategy.setSubProperties` causing problems for non-polymer web components.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/4152)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java, line 182 at r1](https://reviewable.io/reviews/vaadin/flow/4152#-LD6oWUxAgdIbrfJsiJ8:-LD6qGt61Y7fjEDdt8gf:bljh56) ([raw file](https://github.com/vaadin/flow/blob/5f3f7f1c1e39c1a822a4b994956e365a271de488/flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java#L182)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Seems like it's also a no-op in any case where the list isn't directly a child of the element?
</blockquote></details>

Yeap, that's true. I'll include a test for an object that contains a list inside.

---

*[flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java, line 227 at r1](https://reviewable.io/reviews/vaadin/flow/4152#-LD6mrFn1fs5ZSyr7tuR:-LD6pILb10yzm7X3u4yJ:b8qzo9q) ([raw file](https://github.com/vaadin/flow/blob/5f3f7f1c1e39c1a822a4b994956e365a271de488/flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java#L227)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

From my understanding, `UpdatableModelProperties` is about client -> server updates, whereas we're currently processing server -> client updates? What's the connection?
</blockquote></details>

So is this used only on two-way bindings? I'll update the test to have one of the binding as one-way only.

---

*[flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java, line 260 at r1](https://reviewable.io/reviews/vaadin/flow/4152#-LD6o8Sh2ykvdHI2rTyF:-LD6pmqt-KMU3j1Rd6b6:baxydle) ([raw file](https://github.com/vaadin/flow/blob/5f3f7f1c1e39c1a822a4b994956e365a271de488/flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java#L260)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Is there any guarantee that the target element has a `set` function (i.e. that it's an upgraded polymer element)? There was a fix yesterday for the same case with `SimpleElementBindingStrategy.setSubProperties` causing problems for non-polymer web components.
</blockquote></details>

We can add a check to see if the `set` function is available. But what should be the way of setting the property and making the `dom-repeat` (or any other template processing mechanism) work? Would a simple assignment (`element.property = value`) work?

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/4152)*
<!-- Sent from Reviewable.io -->

Legioth commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java, line 227 at r1](https://reviewable.io/reviews/vaadin/flow/4152#-LD6mrFn1fs5ZSyr7tuR:-LD6rw3r9zjbfP0jRrDu:bhzqxff) ([raw file](https://github.com/vaadin/flow/blob/5f3f7f1c1e39c1a822a4b994956e365a271de488/flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java#L227)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

So is this used only on two-way bindings? I'll update the test to have one of the binding as one-way only.
</blockquote></details>

That's how I understand it, but I'm not 100% sure. A test will most likely tell.

---

*[flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java, line 260 at r1](https://reviewable.io/reviews/vaadin/flow/4152#-LD6o8Sh2ykvdHI2rTyF:-LD6s4LnDWD4AHoHbpUf:b-qoa8sj) ([raw file](https://github.com/vaadin/flow/blob/5f3f7f1c1e39c1a822a4b994956e365a271de488/flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java#L260)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

We can add a check to see if the `set` function is available. But what should be the way of setting the property and making the `dom-repeat` (or any other template processing mechanism) work? Would a simple assignment (`element.property = value`) work?
</blockquote></details>

There is no generic way of sending value change notifications for non-polymer components, so all we can do is to put the value in the right place and hope for the best.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/4152)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java, line 182 at r1](https://reviewable.io/reviews/vaadin/flow/4152#-LD6oWUxAgdIbrfJsiJ8:-LD6y9QU3RtUI2rrM8Nl:blf410l) ([raw file](https://github.com/vaadin/flow/blob/5f3f7f1c1e39c1a822a4b994956e365a271de488/flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java#L182)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

Yeap, that's true. I'll include a test for an object that contains a list inside.
</blockquote></details>

Well, apparently having a pojo on the model that contains a List as one of its properties doesn't work at all. The information about the list is never sent to the client-side. So it's a deeper problem, not directly related to the scope of this ticket.

---

*[flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java, line 227 at r1](https://reviewable.io/reviews/vaadin/flow/4152#-LD6mrFn1fs5ZSyr7tuR:-LD74w_g5JBFgL8rBE80:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/5f3f7f1c1e39c1a822a4b994956e365a271de488/flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java#L227)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

That's how I understand it, but I'm not 100% sure. A test will most likely tell.
</blockquote></details>

Done.

---

*[flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java, line 259 at r1](https://reviewable.io/reviews/vaadin/flow/4152#-LD6n_qU0DKjfbqGmnG-:-LD74vHG2Q2KmopbH2lm:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/5f3f7f1c1e39c1a822a4b994956e365a271de488/flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java#L259)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Also (though this also applies to the original code in `SimpleElementBindingStrategy.setSubProperties`), I don't understand why it's necessary to parse the same static JS snippet every time we get here instead of calling a JSNI method that does the same thing.
</blockquote></details>

Done.

---

*[flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java, line 260 at r1](https://reviewable.io/reviews/vaadin/flow/4152#-LD6o8Sh2ykvdHI2rTyF:-LD74tgS5v-CA_rQzIif:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/5f3f7f1c1e39c1a822a4b994956e365a271de488/flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java#L260)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

There is no generic way of sending value change notifications for non-polymer components, so all we can do is to put the value in the right place and hope for the best.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/4152)*
<!-- Sent from Reviewable.io -->

Legioth commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[a discussion](https://reviewable.io/reviews/vaadin/flow/4152#-LDB8_XG2X79lkMuSSLH:-LDB8_XG2X79lkMuSSLI:beymzoc) (no related file):*
What's completely missing here is removal of the old broken logic that always replaced the entire `items` array.

---

*[flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java, line 182 at r1](https://reviewable.io/reviews/vaadin/flow/4152#-LD6oWUxAgdIbrfJsiJ8:-LDB8-IU4xp14KczK2na:b-638ick) ([raw file](https://github.com/vaadin/flow/blob/5f3f7f1c1e39c1a822a4b994956e365a271de488/flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java#L182)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

Well, apparently having a pojo on the model that contains a List as one of its properties doesn't work at all. The information about the list is never sent to the client-side. So it's a deeper problem, not directly related to the scope of this ticket.
</blockquote></details>

That case works for me. Make sure your test code manipulates a proxy from the model instead of the original instance that you put there, e.g. `getModel().getBean().getList().get(0).setFoo("bar")`. If you update the original bean, then Flow won't know that there are any changes.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/4152)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[a discussion](https://reviewable.io/reviews/vaadin/flow/4152#-LDB8_XG2X79lkMuSSLH:-LDMBqYMDYMDWFrUJQbZ:b-prlmg9) (no related file):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

What's completely missing here is removal of the old broken logic that always replaced the entire `items` array.
</blockquote></details>

Added the logic and the test to deal with List of lists, bean with lists, and beans with beans.

---

*[flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java, line 182 at r1](https://reviewable.io/reviews/vaadin/flow/4152#-LD6oWUxAgdIbrfJsiJ8:-LDMC22A7ZX4FNf_LwPE:b-r8i84i) ([raw file](https://github.com/vaadin/flow/blob/5f3f7f1c1e39c1a822a4b994956e365a271de488/flow-client/src/main/java/com/vaadin/client/flow/TreeChangeProcessor.java#L182)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

That case works for me. Make sure your test code manipulates a proxy from the model instead of the original instance that you put there, e.g. `getModel().getBean().getList().get(0).setFoo("bar")`. If you update the original bean, then Flow won't know that there are any changes.
</blockquote></details>

Test updated to also included this scenario.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/4152)*
<!-- Sent from Reviewable.io -->

caalador commented:
Review status: 8 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 172 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_e7zjBe4R0buOBjoJ:-LD_e7zjBe4R0buOBjoK:b1yjzfd) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L172)):*
> ```
> registerPropertyChangesHandlers
> ```
typo -> registerPropertyChangeHandlers

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 282 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_hoU9AzJ67l2AZf15:-LD_hoU9AzJ67l2AZf16:b-lrtk1a) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L282)):*
> ```Java
>                 return false;
>             }
>         } else if (parent.hasFeature(NodeFeatures.ELEMENT_PROPERTIES)) {
> ```

Why do you here switch hasFeature check of TEMPLATE_MODELLIST and ELEMENT_PROPERTIES features.
Even if we can only have one at a time consistency would be nice to have with the checks.

---

*[flow-client/src/test-gwt/java/com/vaadin/client/GwtSuite.java, line 3 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_kRsK2O8qczfgYFe8:-LD_kRsK2O8qczfgYFe9:b8pjhzy) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/test-gwt/java/com/vaadin/client/GwtSuite.java#L3)):*
> ```Java
> package com.vaadin.client;
> 
> import junit.framework.Test;
> ```

No changes except moving imports around?

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/4152)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Review status: 8 of 9 files reviewed at latest revision, 6 unresolved discussions.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 177 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_mLpS0dBR8R6aBBZk:-LD_mLpS0dBR8R6aBBZl:bru39lf) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L177)):*
> ```
> NodeMap map = (NodeMap) feature;
> ```

Why do you cast it here to `NodeMap` ? Is it because it's always a `NodeMap` (then the argument type in the method should be changed).
Or it's because of the check above ?
But this check doesn't say anything about the `feature` . It checks the `node`.
You should either make an `assertion`  or make another check ( e.g. `feature instanceof NodeMap`).

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 184 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_n-XD7G7QDuR2ibR7:-LD_n-XD7G7QDuR2ibR8:bgj6jxh) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L184)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
>                 MapProperty property = event.getProperty();
>                 registrations.push(property.addChangeListener(
>                         change -> handlePropertyChange(property, value)));
>                 handlePropertyChange(property, value);
> ```
</details>


This is hard to read: lambda inside lambda and they are inside `if`.
Please extract somehow.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 188 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_nAoH5Krjs8b6WVGL:-LD_nAoH5Krjs8b6WVGM:bpe0lvo) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L188)):*
> ```
> NodeList list = (NodeList) feature;
> ```

Same here.
either `assert`  or better `if ( node.hasFeature(NodeFeatures.TEMPLATE_MODELLIST) && list == node.getFeature(node.hasFeature(NodeFeatures.TEMPLATE_MODELLIST))`.

Or may be you don't need `feature` argument here at all?

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 195 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_ndXKARIzzIhGeWJw:-LD_ndXKARIzzIhGeWJx:b-tou5un) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L195)):*
> ```
> addUnregisterListener(
> ```

This one also returns a registration which needs to be removed  :)

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 230 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_o25LDuaL5YSKEIRz:-LD_o25LDuaL5YSKEIS-:b-ibgjom) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L230)):*
<details><summary><i alt="summary">Quoted 12 lines of code…</i></summary>

> ```
>             JsArray<String> path = JsCollections.array();
>             if (buildNotificationPath(root, node, path)) {
>                 StringBuilder pathBuilder = new StringBuilder();
>                 String sep = "";
>                 for (int i = path.length() - 1; i >= 0; i--) {
>                     pathBuilder.append(sep).append(path.get(i));
>                     sep = ".";
>                 }
>                 splice((Element) root.getDomNode(), pathBuilder.toString(),
>                         index, remove, WidgetUtil.crazyJsoCast(array));
>                 return;
>             }
> ```
</details>


Too many nested blocks: `for` inside `if` inside `if`. Please refactor.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 244 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_oYPb076504EwGohG:-LD_oYPb076504EwGohH:b1x6rkr) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L244)):*
> ```
>  private static void doHandlePropertyChange(
> ```

The code in this method looks quite similar to the code above.
Is it possible to reuse some parts from there/extract duplicated code?

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 274 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_op3oCnsRv0GG8AcZ:-LD_op3oCnsRv0GG8Ac_:b-tezk2x) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L274)):*
> ```
> buildNotificationPath
> ```

The method name doesn't tell me what does it return.
It should return some `boolean` but I don't really understand what this boolean means.

As I understand this is an additional value to the work done with the `path`.
Shouldn't it just return the `path` is its return value is suggested to be `true` and e.g. `null` otherwise ?

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 294 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_p2V9DcEaPn20p-DI:-LD_p2V9DcEaPn20p-DJ:bh4oyrn) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L294)):*
> ```
> buildListNotificationPath
> ```

Same issue with the method name here.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 315 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_p7MZBTTVdgQUMe0V:-LD_p7MZBTTVdgQUMe0W:b-xh5fc6) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L315)):*
> ```
> buildPropertiesNotificationPath
> ```


And again : the method name.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 344 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_q94Y2NyUGfOyc0OQ:-LD_q94Y2NyUGfOyc0OR:b-jse2l5) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L344)):*
> ```
> getParentNodeAttachedToADomNode
> ```

The method name is too long and `A` is extra.
`getParentWithDOMNode`(may be).
Though it's hard to invent good name here :(

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/4152)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[flow-client/src/test-gwt/java/com/vaadin/client/flow/GwtPolymerModelTest.java, line 354 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_vQBwDVwHk_bcStMI:-LD_vQBwDVwHk_bcStMJ:beerj2l) ([raw file](https://github.com/vaadin/flow/blob/d39f957217a96528caaa0aa5e6b65b5132c14719/flow-client/src/test-gwt/java/com/vaadin/client/flow/GwtPolymerModelTest.java#L354)):*
> ```
> testUpdateModelSubProperty_subpropertyIsUpdatableAndIsNotSetFromServer_subpropertyIsSyn
> ```

No replacement ?

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/4152)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 172 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_e7zjBe4R0buOBjoJ:-LD_nfA325KC5uCIp33y:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L172)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

> ```
> registerPropertyChangesHandlers
> ```
typo -> registerPropertyChangeHandlers
</blockquote></details>

Done.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 177 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_mLpS0dBR8R6aBBZk:-LD_r_AQAGR_bqR60RR-:bh7a1bw) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L177)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
> NodeMap map = (NodeMap) feature;
> ```

Why do you cast it here to `NodeMap` ? Is it because it's always a `NodeMap` (then the argument type in the method should be changed).
Or it's because of the check above ?
But this check doesn't say anything about the `feature` . It checks the `node`.
You should either make an `assertion`  or make another check ( e.g. `feature instanceof NodeMap`).
</blockquote></details>

Added an assert, since the checks are done previously in the chain, and adding another `if` would be redundant.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 184 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_n-XD7G7QDuR2ibR7:-LD_tWTb0CRbcrQExYED:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L184)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
>                 MapProperty property = event.getProperty();
>                 registrations.push(property.addChangeListener(
>                         change -> handlePropertyChange(property, value)));
>                 handlePropertyChange(property, value);
> ```

This is hard to read: lambda inside lambda and they are inside `if`.
Please extract somehow.
</blockquote></details>

Done.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 188 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_nAoH5Krjs8b6WVGL:-LD_rMW69mDYd7nBfel9:bj9cwx1) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L188)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
> NodeList list = (NodeList) feature;
> ```

Same here.
either `assert`  or better `if ( node.hasFeature(NodeFeatures.TEMPLATE_MODELLIST) && list == node.getFeature(node.hasFeature(NodeFeatures.TEMPLATE_MODELLIST))`.

Or may be you don't need `feature` argument here at all?
</blockquote></details>

The list is needed down bellow, to add the `SpliceListener`.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 195 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_ndXKARIzzIhGeWJw:-LD_uJKR7C4oT1hgwufG:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L195)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
> addUnregisterListener(
> ```

This one also returns a registration which needs to be removed  :)
</blockquote></details>

Done.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 230 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_o25LDuaL5YSKEIRz:-LDa4EFm1gLR3G5ECR4m:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L230)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
>             JsArray<String> path = JsCollections.array();
>             if (buildNotificationPath(root, node, path)) {
>                 StringBuilder pathBuilder = new StringBuilder();
>                 String sep = "";
>                 for (int i = path.length() - 1; i >= 0; i--) {
>                     pathBuilder.append(sep).append(path.get(i));
>                     sep = ".";
>                 }
>                 splice((Element) root.getDomNode(), pathBuilder.toString(),
>                         index, remove, WidgetUtil.crazyJsoCast(array));
>                 return;
>             }
> ```

Too many nested blocks: `for` inside `if` inside `if`. Please refactor.
</blockquote></details>

Done.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 244 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_oYPb076504EwGohG:-LDa3dAf8XHBv8PGnLkF:bahqkzo) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L244)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
>  private static void doHandlePropertyChange(
> ```

The code in this method looks quite similar to the code above.
Is it possible to reuse some parts from there/extract duplicated code?
</blockquote></details>

The structure is similar, but since the List requires more stuff for the splice, such as indexes and number of items to remove and add, making this code generic enough to handle both cases would require lots of inner ifs or lambdas everywhere, and in the end it would be more confusing than in the current form.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 274 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_op3oCnsRv0GG8AcZ:-LD_uWkS87_YeDsRFars:br9guls) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L274)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
> buildNotificationPath
> ```

The method name doesn't tell me what does it return.
It should return some `boolean` but I don't really understand what this boolean means.

As I understand this is an additional value to the work done with the `path`.
Shouldn't it just return the `path` is its return value is suggested to be `true` and e.g. `null` otherwise ?
</blockquote></details>

Changed the method to make it clearer. Since it is a recursive method, using the flag was easier in the first run.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 294 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_p2V9DcEaPn20p-DI:-LD_yFOZ4BEsJIoHCaha:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L294)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
> buildListNotificationPath
> ```

Same issue with the method name here.
</blockquote></details>

Done.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 315 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_p7MZBTTVdgQUMe0V:-LD_yG1666PY-EUTgSf0:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L315)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
> buildPropertiesNotificationPath
> ```


And again : the method name.
</blockquote></details>

Done.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 344 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_q94Y2NyUGfOyc0OQ:-LD_yT2uB27skYn9A-ix:b-g94o95) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L344)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
> getParentNodeAttachedToADomNode
> ```

The method name is too long and `A` is extra.
`getParentWithDOMNode`(may be).
Though it's hard to invent good name here :(
</blockquote></details>

Changed it, see if it is better.

---

*[flow-client/src/test-gwt/java/com/vaadin/client/GwtSuite.java, line 3 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_kRsK2O8qczfgYFe8:-LD_nVvf91Ewn_KoxL8S:b4lfpk0) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/test-gwt/java/com/vaadin/client/GwtSuite.java#L3)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

No changes except moving imports around?
</blockquote></details>

Yeah, we changed and reverted the changes on the class during the tests and Eclipse changed the imports.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/4152)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: 7 of 8 files reviewed at latest revision, 17 unresolved discussions.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 282 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_hoU9AzJ67l2AZf15:-LDaHrW46bs2yxap8u5h:b4fesn3) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L282)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

Why do you here switch hasFeature check of TEMPLATE_MODELLIST and ELEMENT_PROPERTIES features.
Even if we can only have one at a time consistency would be nice to have with the checks.
</blockquote></details>

What do you mean? Creating a separate method for each of them?

---

*[flow-client/src/test-gwt/java/com/vaadin/client/flow/GwtPolymerModelTest.java, line 354 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_vQBwDVwHk_bcStMI:-LDaIXnjBmKP-KXHL42O:b-733a9p) ([raw file](https://github.com/vaadin/flow/blob/d39f957217a96528caaa0aa5e6b65b5132c14719/flow-client/src/test-gwt/java/com/vaadin/client/flow/GwtPolymerModelTest.java#L354)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
> testUpdateModelSubProperty_subpropertyIsUpdatableAndIsNotSetFromServer_subpropertyIsSyn
> ```

No replacement ?
</blockquote></details>

According to @legioth this test was about something we don't actually support, so it wouldn't make sense to keep and patch it.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/4152)*
<!-- Sent from Reviewable.io -->

caalador commented:
Review status: 7 of 8 files reviewed at latest revision, 16 unresolved discussions.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 282 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_hoU9AzJ67l2AZf15:-LDaKlBz1np4EPpL79WR:b-iotxxh) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L282)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

What do you mean? Creating a separate method for each of them?
</blockquote></details>

No I meant that in polymerUtils for example you always check `hasFeature(ELEMENT_PROPERTIES)` first and then `

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/4152)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: 7 of 8 files reviewed at latest revision, 15 unresolved discussions.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 282 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_hoU9AzJ67l2AZf15:-LDaNsp73gDkpNKLWf5M:bkrd32s) ([raw file](https://github.com/vaadin/flow/blob/eb3cfc0e083d76e6ba444393f55e8242581b8ba8/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L282)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

No I meant that in polymerUtils for example you always check `hasFeature(ELEMENT_PROPERTIES)` first and then `
</blockquote></details>

I see. Ok, changed the order.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/4152)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: 6 of 8 files reviewed at latest revision, 6 unresolved discussions.

---

*[flow-client/src/main/java/com/vaadin/client/PolymerUtils.java, line 207 at r6](https://reviewable.io/reviews/vaadin/flow/4152#-LDaSFIb97ytU_TQt5py:-LDaktoL5rFvCLAZIS6t:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/1f14e49bec0766374c9369aafe4cbbb42be37c7a/flow-client/src/main/java/com/vaadin/client/PolymerUtils.java#L207)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> ```
> registerPropertyChangeHandlerForEachProperty
> ```

I would say `ForEachProperty` is an extra.
</blockquote></details>

Done.

---

*[flow-client/src/test-gwt/java/com/vaadin/client/flow/GwtPolymerModelTest.java, line 354 at r4](https://reviewable.io/reviews/vaadin/flow/4152#-LD_vQBwDVwHk_bcStMI:-LDakuob-Nk1aNKaVNOm:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/d39f957217a96528caaa0aa5e6b65b5132c14719/flow-client/src/test-gwt/java/com/vaadin/client/flow/GwtPolymerModelTest.java#L354)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

I don't think so.

This method tests that if the model has subproperty which is allowed to be updatable from the client side to the server side then its changes are sent to the server.

If we don't support it then it means that `@AllowClientUpdates` doesn't work for subproperties .
I would expect that it works actually.
And this code should not have any relation to lists. So I don't really see a reason why it should be removed (or seriously broken) because of changes in this PR.

What we don't support at the moment is : changes in bean properties inside a list.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/4152)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Review status: 7 of 8 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/4152)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/4152#-:-LDamHkwDNIaOM0J9qbh:bnfp4nl)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/flow/4152#-:-LDap9LQ7GsMTsf76fwX:bnfp4nl)*
<!-- Sent from Reviewable.io -->
